### DISCUSSION TITLE
Preventing database loss on network mount failure

### DISCUSSION THREAD

**Discussion by Scrath1** [navidrome/navidrome#3033] [Ideas]
Hello everyone,
I recently ran into an issue where my database was lost following a power failure. The reason for this was not data corruption but rather that upon restarting of my servers the network share on which my music is stored failed to mount. When navidrome started up and ran its first media scan it detected that the music was "removed" and tossed all tracks from its database causing me to lose my song ratings, favorites and playlists.

I am aware that this is something that can and probably should be prevented on the layer of the OS itself but it would be nice to have another failsafe in the application for scenarios like this. For now I added the successful mounting of the network share to the requirements for the docker service to start up.

What I would propose as a failsafe in navidrome itself is that when the application detects a number of songs above a configurable threshold have been removed from the filesystem, a backup of the database is created first before removing those songs.

Edit:
For those interested in making docker require a successful network mount, [this](https://unix.stackexchange.com/questions/279291/systemd-how-to-require-and-effect-mounting-a-specific-file-system) is what I did. Note the comment to the accepted reply which contains a command to list the required mount point name. I added this name to the systemd service file for docker.

**Comment by metalheim** [comment]
It's certainly not nice if this happens. 
There actually is a failsave in the navidrome scanner already that checks if there are 0 files and 0 subfolders in your library and abort the scan (thus no music should be removed).
That logic is here:
https://github.com/navidrome/navidrome/blob/a7a4fb522c5221d502ec73901375f59d2871d009/scanner/tag_scanner.go#L87
Can you please check if there were indeed 0 files and 0 subfolders in your MusicFolder when this happened to you?

**That said:**
The core issue (files are not there anymore) is best handled on the service-level, so that navidrome istnt even started if the music isn't there.
Also, a backup of the database should be made regularly, so that all playlists etc. can be re-imported in case of a catastrophic database wipe.

**Reply by Scrath1** [reply]
I'm not sure how but I believe a couple songs somehow remained since I had about 3 of my 200 favorites left. Or maybe what I believe happened didn't actually happen this way. Nice to see there is actually a failsafe even though it didn't help in my case. I'm unfortunately already in the process of restoring my favorites from an offline cache that my app on my phone uses and have remounted the server yesterday so I'm not sure how the directory really looked at the moment it failed.

I agree with the service-level handling which I believe I implemented correctly on my setup now.

Regarding the backup of the database, I thought this would be quite complicated as I would have to shutdown the container, backup the database and restart the container at certain timeintervals. It wasn't until I was reading through the forum in regards to my current issue that I found out people have had little issue with simply copying the database files while the container is running.
